### PR TITLE
Framework additional

### PR DESCRIPTION
**Pull request by dfedyk**
Some additional changes from Routing Director Review

### CODE REVIEW

File: data-plane-framework/draft-ietf-detnet-data-plane-framework.xml [hunk 1]
```diff
@@ -224,77 +232,88 @@
...
+         connectivity function of the forwarding sub-layer. An example of this
+         is Packet Replication, Elimination, and Ordering functions see <xref
+                 target="PREOF"/>.  The ordering (POF) uses sequence numbers
+         added to packets enabling a range of packet order protection from
+         simple ordering and dropping out-of-order packets to more complex
+         reordering of a fixed number of out-of-order, minimally delayed
+         packets.  Reordering requires buffer resources and has impact on the
+	 delay and jitter of packets in the DetNet flow. 
+	 These functions are optional and many require sequence numbers that are 
+	 provided by the service sub-layer or available to the service sub-layer from a higher layer
```
Comment (vb-balazs) [inline data-plane-framework/draft-ietf-detnet-data-plane-framework.xml:263]: "available to the service sub-layer from a higher layer" no such option in current data plane drafts.

File: data-plane-framework/draft-ietf-detnet-data-plane-framework.xml [hunk 2]
```diff
@@ -635,6 +657,22 @@ CE1----EN1--------R1-------R2-------R3--------EN2-----CE2
...
+     In the preceding example, reordering of packets is dependent on the number
+     of out-of-order packets that can be buffered and the delay difference of
+     arriving packets.  DetNet uses configuration of the number of maximum
+     number out-of-order packets and the maximum delay that out-of-order
+     packets can be held before being delivered. If the differential delay
+     between paths is large enough or the arrival rate of packets large enough
+     packets may be dropped instead of being reordered.  Likewise elimination
+     uses the sequence number to eliminate duplicate packets and large
+     differential delays combined with high numbers of packets may exceed the
+     ability of this elimination function to eliminate all duplicate packets.
```
Comment (vb-balazs) [inline data-plane-framework/draft-ietf-detnet-data-plane-framework.xml:672]: I think this is not valid. PEF works with a "seg.num window" and a "history". Packets with Seq.num out of the "window" are dropped. "Window" moves according to the highest received seq.num.  
Therefore PEF cannot result in scenarios where "... may exceed the ability of this elimination function to eliminate all duplicate packets". The worst case of PEF is to discard packets that should not be discarded.
Design of PEF parameters requires that "window" parameter is big enough to deal with the differential delay.

File: data-plane-framework/draft-ietf-detnet-data-plane-framework.xml [hunk 3]
```diff
@@ -635,6 +657,22 @@ CE1----EN1--------R1-------R2-------R3--------EN2-----CE2
...
      established and maintained for the duration of the DetNet service.
      </t>
+   </section>
+     <section title="Path Differential Delay">
+             <t>
+     In the preceding example, reordering of packets is dependent on the number
+     of out-of-order packets that can be buffered and the delay difference of
+     arriving packets.  DetNet uses configuration of the number of maximum
+     number out-of-order packets and the maximum delay that out-of-order
+     packets can be held before being delivered. If the differential delay
```
Comment (vb-balazs) [inline data-plane-framework/draft-ietf-detnet-data-plane-framework.xml:667]: Maximum delay is a parameter for the DetNet flow and not for the delay of "out-of-order" packets. However they are somewhat related. "Maximum delay" decreased by the "actual delay of the packet" gives the delay what can be used during re-ordering.

File: data-plane-framework/draft-ietf-detnet-data-plane-framework.xml [hunk 4]
```diff
@@ -635,6 +657,22 @@ CE1----EN1--------R1-------R2-------R3--------EN2-----CE2
...
+     between paths is large enough or the arrival rate of packets large enough
+     packets may be dropped instead of being reordered.  Likewise elimination
```
Comment (vb-balazs) [inline data-plane-framework/draft-ietf-detnet-data-plane-framework.xml:669]: I think this is not precise. What this text says is that if POF buffer is too small (bad design!) then packet drops occur, what is trivial.
The prerequisite for POF is that the "differential delay" must be smaller than the "Maximum delay" minus the "fastest path delay of the packet". If this is fulfilled than POF buffer can be calculated. Arrival rate does matter from re-ordering perspective only for calculating how big buffers are needed.

**Comment by dfedyk** [comment]
HI Balázs 

 

Comments [Don]. 

 

Cheers,

Don

File: data-plane-framework/draft-ietf-detnet-data-plane-framework.xml (hunk 1 above)
Comment (vb-balazs) [reply data-plane-framework/draft-ietf-detnet-data-plane-framework.xml:263]: [Don] This is straight from the DetNet Architecture.  The question was is Sequence number available to IP? The general answer is not at IP but maybe with an option or at a higher layer. The framework options do not have to apply only to the data planes we have today.

[Balazs] Ack. No change needed.

File: data-plane-framework/draft-ietf-detnet-data-plane-framework.xml (hunk 2 above)
Comment (vb-balazs) [reply data-plane-framework/draft-ietf-detnet-data-plane-framework.xml:672]: [Don] A lot of this text comes from the TSN documents.  I checked what TSN says.   And in high speed scenarios a sequence number of 16 bits is not very long.  I think there are valid duplicate conditions.

[Balazs] This part of the TSN docu is not well written. The major problem of PEF in the here referred case is that slow path packets are dropped (they are out-of-window). Duplicate delivery may occur in very-very special cases. So proposed change:
"... may exceed the ability of this elimination function to work properly."

File: data-plane-framework/draft-ietf-detnet-data-plane-framework.xml (hunk 3 above)
Comment (vb-balazs) [reply data-plane-framework/draft-ietf-detnet-data-plane-framework.xml:667]: [Don] Please suggest corrected text.

[Balazs] What about something like this:
"DetNet uses for configuration of POF related buffers 
the DetNet flow specific requirements (e.g., maximum
number of out-of-order packets, maximum latency 
of the DetNet flow, etc.)."

File: data-plane-framework/draft-ietf-detnet-data-plane-framework.xml (hunk 4 above)
Comment (vb-balazs) [reply data-plane-framework/draft-ietf-detnet-data-plane-framework.xml:669]: [Don] Again this is TSN derived text. ...

[Balazs] In that part of CB text is broken. :--((( CB does not really define details of a POF.

**Review by vb-balazs — COMMENTED** [review]
Proposed re-write of "3.6.2.2. Path Differential Delay"

"In the preceding example, proper working of duplicate 
elimination and reordering of packets are dependent 
on the number of out-of-order packets that can be 
buffered and the delay difference of arriving packets. 
DetNet uses for configuration of PEF and POF related 
buffers the DetNet flow specific requirements (e.g., 
maximum number of out-of-order packets, maximum 
latency of the DetNet flow, etc.)."

If the differential delay between paths is large enough 
or the arrival rate of packets large enough packets may 
be dropped instead of being reordered. Path selection 
for DetNet services needs to take these factors into 
account."

**Pull request closed without merging**